src/uct/sm/mm/mm_iface.c

```diff
@@ -95,20 +133,209 @@ static uct_iface_ops_t uct_mm_iface_ops = {
     .ep_destroy          = UCS_CLASS_DELETE_FUNC_NAME(uct_mm_ep_t),
 };
 
+static inline void uct_mm_progress_fifo_tail(uct_mm_iface_t *iface)
+{
+    /* don't progress the tail every time - release in batches. improves performance */
+    if (iface->read_index & iface->fifo_release_factor_mask) return;
+
+    while ((iface->recv_fifo_ctl->tail < iface->read_index)) {
+        uct_mm_flush();
+        iface->recv_fifo_ctl->tail++;
+    }
+}
+
+static void uct_mm_iface_process_recv(uct_mm_iface_t *iface,
+                                      uct_mm_fifo_element_t* elem_to_read)
+{
+    if (elem_to_read->flags & UCT_MM_FIFO_ELEM_FLAG_INLINE) {
+        /* read inline messages from the FIFO elements */
+        uct_mm_iface_invoke_am(iface, elem_to_read->am_id, (elem_to_read + 1),
+                               elem_to_read->length, iface->last_recv_desc);
+    } else {
+        ucs_fatal("bcopy is not implemented");
+    }
+}
+
+static inline void uct_mm_iface_poll_fifo(uct_mm_iface_t *iface)
+{
+    uint64_t read_index_loc, read_index;
+    uct_mm_fifo_element_t* read_index_elem;
+
+    read_index = iface->read_index;
+    read_index_loc = (read_index & iface->fifo_mask);
+    /* the fifo_element which the read_index points to */
+    read_index_elem = (uct_mm_fifo_element_t*) ((char*) iface->recv_fifo_elements +
+                      (read_index_loc * (iface->elem_size)));
+
+    /* check the read_index to see if there is a new item to read (checking the owner bit) */
+    if (((read_index >> iface->fifo_shift) & 1)	== ((read_index_elem->flags) & 1)) {
+        /* raise the read_index. */
+        iface->read_index++;
+
+        /* could progress the read_index. now read from read_index_elem */
+        ucs_memory_cpu_load_fence();
+        ucs_assert(iface->read_index <= iface->recv_fifo_ctl->head);
+
+        uct_mm_iface_process_recv(iface, read_index_elem);
+        uct_mm_progress_fifo_tail(iface);
```

Comment:
i will check. but for now we would only know if it affects perf for only two processes. since all senders send to the same fifo (and therefore see the same tail), it might affect the performance when there are many ranks running on a host.